File: common/src/main/java/com/github/ykiselev/common/memory/pool/ObjectPoolFrame.java
package com.github.ykiselev.common.memory.pool;

import java.util.ArrayList;
import java.util.List;

import static java.util.Objects.requireNonNull;

public final class ObjectPoolFrame<T> implements AutoCloseable {

    private final ObjectPool<T> owner;

    private final List<T> pool = new ArrayList<>();

    private int index;

    int size() {
        return pool.size();
    }

    int index() {
        return index;
    }

    ObjectPoolFrame(ObjectPool<T> owner) {
        this.owner = requireNonNull(owner);
    }

    public T allocate() {
        if (index == pool.size()) {
            pool.add(owner.newInstance());
        }
        return pool.get(index++);
    }

    @Override
    public void close() {
        index = 0;
        owner.pop(this);
    }
}

File: common/src/main/java/com/github/ykiselev/common/memory/pool/ObjectPoolFrame.kt
package com.github.ykiselev.common.memory.pool

import com.github.ykiselev.common.memory.Allocator
import java.util.*

class ObjectPoolFrame<T> internal constructor(val owner: ObjectPool<T>) : AutoCloseable, Allocator<T> {

    private val pool: MutableList<T> = ArrayList()

    private var index = 0

    fun size(): Int {
        return pool.size
    }

    fun index(): Int {
        return index
    }

    override fun allocate(): T {
        if (index == pool.size) {
            pool.add(owner.newInstance())
        }
        return pool[index++]
    }

    override fun close() {
        index = 0
        owner.pop(this)
    }
}
